src/main/java/com/zendesk/maxwell/producer/MaxwellKinesisProducer.java

```diff
@@ -32,82 +31,169 @@
+public class MaxwellKinesisProducer extends AbstractProducer {
+	private final ArrayBlockingQueue<RowMap> queue;
+	private final MaxwellKinesisProducerWorker worker;
+
+	public MaxwellKinesisProducer(MaxwellContext context, String kinesisStream) {
+		super(context);
+		this.queue = new ArrayBlockingQueue<>(100);
+		this.worker = new MaxwellKinesisProducerWorker(context, kinesisStream, this.queue);
+		new Thread(this.worker, "maxwell-kinesis-worker").start();
+	}
+
+	@Override
+	public void push(RowMap r) throws Exception {
+		this.queue.put(r);
+	}
+
+}
+
+class MaxwellKinesisProducerWorker extends AbstractAsyncProducerWorker {
+
 	private static final Logger logger = LoggerFactory.getLogger(MaxwellKinesisProducer.class);
 
 	private final MaxwellKinesisPartitioner partitioner;
 	private final KinesisProducer kinesisProducer;
 	private final String kinesisStream;
-
-	public MaxwellKinesisProducer(MaxwellContext context, String kinesisStream) {
-		super(context);
-
+	private final int maxAttempts;
+	private final int maxBufferedRecords;
+	private AtomicInteger successRecords = new AtomicInteger(0);
+	private AtomicInteger errorRecords = new AtomicInteger(0);
+	private final ArrayBlockingQueue<RowMap> queue;
+
+	public MaxwellKinesisProducerWorker(MaxwellContext context, String kinesisStream, ArrayBlockingQueue<RowMap> queue) {
+		super(context, queue);
 		String partitionKey = context.getConfig().producerPartitionKey;
 		String partitionColumns = context.getConfig().producerPartitionColumns;
 		String partitionFallback = context.getConfig().producerPartitionFallback;
 		boolean kinesisMd5Keys = context.getConfig().kinesisMd5Keys;
 		this.partitioner = new MaxwellKinesisPartitioner(partitionKey, partitionColumns, partitionFallback, kinesisMd5Keys);
 		this.kinesisStream = kinesisStream;
 
+		this.maxAttempts = context.getConfig().kinesisMaxAttempts;
+		this.maxBufferedRecords = context.getConfig().kinesisMaxBufferedRecords;
+
 		Path path = Paths.get("kinesis-producer-library.properties");
 		if(Files.exists(path) && Files.isRegularFile(path)) {
 			KinesisProducerConfiguration config = KinesisProducerConfiguration.fromPropertiesFile(path.toString());
 			this.kinesisProducer = new KinesisProducer(config);
 		} else {
 			this.kinesisProducer = new KinesisProducer();
 		}
+		this.queue = queue;
 	}
 
 	@Override
 	public void sendAsync(RowMap r, AbstractAsyncProducer.CallbackCompleter cc) throws Exception {
 		String key = this.partitioner.getKinesisKey(r);
 		String value = r.toJSON(outputConfig);
 
-		ByteBuffer encodedValue = ByteBuffer.wrap(value.getBytes("UTF-8"));
-		ListenableFuture<UserRecordResult> future = kinesisProducer.addUserRecord(kinesisStream, key, encodedValue);
-
-		// release the reference to ease memory pressure
-		if(!KinesisCallback.logger.isDebugEnabled()) {
-			value = null;
+		// Wait for KPL to finish sending events
+		while(kinesisProducer.getOutstandingRecordsCount()> this.maxBufferedRecords) {
+			Thread.sleep(1);
```

Comment:
This is for avoiding backpressure when Kinesis can not handle the events at the same speed you produce them. KPL has an internal buffer but if It grows infinitely It can consume your memory or events can reach their ttl.